benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90

```diff
@@ -0,0 +1,668 @@
+! Constructors
+!-------------------------------------------------------------------------------
+
+! overload the default structure constructor for function space
+interface matrix_vector_kernel_type
+   module procedure matrix_vector_kernel_constructor
+end interface
+
+!-------------------------------------------------------------------------------
+! Contained functions/subroutines
+!-------------------------------------------------------------------------------
+!public matrix_vector_code
+contains
+
+  type(matrix_vector_kernel_type) function matrix_vector_kernel_constructor() result(self)
+  return
+end function matrix_vector_kernel_constructor
+
+!> @brief Computes lhs = matrix*x
+!> @param[in] cell Horizontal cell index
+!! @param[in] nlayers Number of layers
+!! @param[in] ncell_3d Total number of cells
+!! @param[in] ndf1 Number of degrees of freedom per cell for the output field
+!! @param[in] undf1 Unique number of degrees of freedom  for the output field
+!! @param[in] map1 Dofmap for the cell at the base of the column for the output field
+!! @param[in] map2 Dofmap for the cell at the base of the column for the input field
+!! @param[in] ndf2 Number of degrees of freedom per cell for the input field
+!! @param[in] undf2 Unique number of degrees of freedom for the input field 
+!! @param[in] x Input data
+!> @param[inout] lhs Output lhs (A*x)
+!! @param[in] matrix Local matrix assembly form of the operator A 
+subroutine matrix_vector_code_original(cell,        &
+                              nlayers,     &
+                              lhs, x,      & 
+                              ncell_3d,    &
+                              matrix,      &
+                              ndf1, undf1, map1, &
+                              ndf2, undf2, map2)
+ 
+  !Arguments
+  integer,                   intent(in)    :: cell, nlayers, ncell_3d
+  integer,                   intent(in)    :: undf1, ndf1
+  integer,                   intent(in)    :: undf2, ndf2
+  integer, dimension(ndf1),  intent(in)    :: map1
+  integer, dimension(ndf2),  intent(in)    :: map2
+  real(kind=r_def), dimension(undf2),              intent(in)    :: x
+  real(kind=r_def), dimension(undf1),              intent(inout) :: lhs
+  real(kind=r_def), dimension(ndf1,ndf2,ncell_3d), intent(in)    :: matrix
+
+  !Internal variables
+  integer                           :: df, k, ik , df2
+  real(kind=r_def), dimension(ndf2) :: x_e
+  real(kind=r_def), dimension(ndf1) :: lhs_e
+ 
+  do k = 0, nlayers-1
+    do df = 1, ndf2  
+       x_e(df) = x(map2(df)+k)
+    end do
+
+    do df = 1, ndf1
+       lhs_e(df) = lhs(map1(df)+k)
+    end do
+
+    ik = (cell-1)*nlayers + k + 1
+
+    lhs_e = matmul(matrix(:,:,ik),x_e)
+    
+    do df = 1,ndf1
+       lhs(map1(df)+k) = lhs(map1(df)+k) + lhs_e(df) 
+    end do
+ end do
+ 
+end subroutine matrix_vector_code_original
+
+subroutine matrix_vector_code_fuseinterleaved(cell,        &
```

Comment:
Please could you add a comment to each of the various implementations that briefly explains what it is that they test/do.

Reply:
Done

Reply:
Done